compiler/src/main/java/com/squareup/anvil/compiler/codegen/incremental/GeneratedFileCache.kt

```diff
@@ -0,0 +1,209 @@
+
+      check(!hasCycle) {
+        """
+          Adding this mapping would create a cycle.
+             source path: ${source.absolute(projectDir).file}
+          generated path: ${generated.file}
+        """.trimIndent()
+      }
+
+      sourcesToGenerated.add(source, generatedRelative)
+      generatedToSources.add(generatedRelative, source)
+      filesToMd5.computeIfAbsent(source) { source.md5() }
+    }
+  }
+
+  fun removeSource(sourceFile: SourceFile) {
+    when (sourceFile) {
+      is SourceAbsoluteFile -> removeSource(sourceFile.relativeTo(projectDir))
+      is SourceRelativeFile -> {
+        filesToMd5.remove(sourceFile)
+        generatedToSources.remove(sourceFile.asGeneratedFile())
+
+        // All generated files that claim this source file as a source.
+        val generated = sourcesToGenerated.remove(sourceFile) ?: return
+
+        for (gen in generated) {
+
+          // For any generated file that has multiple sources,
+          // remove that generated file from the "generated" set for those other sources.
+          // Note that this does not call `removeSource` for that other source.
+          for (otherSource in generatedToSources[gen]) {
+            if (otherSource != sourceFile) {
+              sourcesToGenerated.remove(otherSource, gen)
+            }
+          }
+          removeSource(gen.asSourceFile())
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns `true` if the [sourceFile] has changed since the last time it was added to the cache.
+   */
+  fun hasChanged(sourceFile: SourceFile): Boolean {
+    return when (sourceFile) {
+      is SourceAbsoluteFile -> hasChanged(sourceFile.relativeTo(projectDir))
+      is SourceRelativeFile -> {
+        if (sourceFile == ANY) return true
+        val previousMd5 = filesToMd5[sourceFile] ?: return true
+        val currentMd5 = sourceFile.md5()
+        !currentMd5.contentEquals(previousMd5)
+      }
+    }
+  }
+
+  override fun close() {
+    writeToBinaryFile()
+  }
+
+  private fun RelativeFile.md5(): ByteArray {
+    if (this == ANY) return ByteArray(0)
+    val absoluteFile = projectDir.resolve(this).file
+    return MessageDigest.getInstance("MD5")
+      .digest(absoluteFile.readBytes())
+  }
+
+  private fun ByteArray.hex() = joinToString("") { "%02x".format(it) }
+
+  override fun toString(): String = """
+    |======================== ${this::class.simpleName}
+    | -- sourcesToGenerated
+    |$sourcesToGenerated
+    |
+    | -- generatedToSources
+    |$generatedToSources
+    |
+    | -- filesToMd5
+    |${filesToMd5.toList().joinToString("\n") { "${it.first} -> ${it.second.hex()}" }}
+    |========================
+  """.trimMargin()
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (other !is GeneratedFileCache) return false
+
+    if (generatedToSources != other.generatedToSources) return false
+    if (sourcesToGenerated != other.sourcesToGenerated) return false
```

Comment:
Should we be checking `generatedToContent` here too?

Reply:
Updated